### PR TITLE
Migrate cluster metadata store to use cluster metadata info table

### PR DESCRIPTION
**Pull request by yux0**
**What changed?**
1. Migrate cluster metadata store to use cluster metadata info table
2. Wire cluster metadata persistence api with the new table
3. Add delete cluster metadata query
4. Update fx backfill cluster metadata logic
5. Update Makefile

**Why?**
To dynamically update cluster metadata

**How did you test it?**
Integration tests

**Potential risks**
The check version API cannot find the version info from the cluster metadata

**Is hotfix candidate?**
N/A

### CODE REVIEW

File: common/persistence/cassandra/cluster_metadata_store.go [hunk 1]
```diff
@@ -130,6 +137,14 @@ func (m *ClusterMetadataStore) SaveClusterMetadata(request *p.InternalSaveCluste
 	return true, nil
 }
 
+func (m *ClusterMetadataStore) DeleteClusterMetadata(request *p.InternalDeleteClusterMetadataRequest) error {
+	query := m.session.Query(templateDeleteClusterMetadata, constMetadataPartition, request.ClusterName)
```
Comment (yiminc) [inline common/persistence/cassandra/cluster_metadata_store.go:141]: what would happen if current cluster is deleted from db?
should we prevent deleting current cluster?

File: temporal/fx.go [hunk 2]
```diff
@@ -481,46 +481,94 @@ func ApplyClusterMetadataConfigProvider(
...
+			// Only set current cluster Id as we don't know the remote cluster Id.
+			clusterId = uuid.New()
+		}
+		// We assume the existing remote cluster info is correct.
+		applied, err := clusterMetadataManager.SaveClusterMetadata(
+			&persistence.SaveClusterMetadataRequest{
+				ClusterMetadata: persistencespb.ClusterMetadata{
+					HistoryShardCount:        config.Persistence.NumHistoryShards,
+					ClusterName:              clusterName,
+					ClusterId:                clusterId,
```
Comment (yiminc) [inline temporal/fx.go:497]: but we don't have the clusterId from existing config for remote clusters, right?

File: temporal/fx.go [hunk 3]
```diff
@@ -481,46 +481,94 @@ func ApplyClusterMetadataConfigProvider(
...
+		if err != nil {
+			return config.ClusterMetadata,
+				config.Persistence,
+				fmt.Errorf("error while fetching metadata from cluster %s: %w", clusterName, err)
+		}
 
-		config.ClusterMetadata.CurrentClusterName = resp.ClusterName
-	}
+		if clusterName == clusterData.CurrentClusterName {
+			if config.ClusterMetadata.CurrentClusterName != resp.ClusterName {
```
Comment (yiminc) [inline temporal/fx.go:521]: how could this happen? the resp is from GetCurrentClusterMetadata() which load from DB for currentClusterName.

Comment (yiminc) [reply temporal/fx.go:521]: If the current cluster name in config file is changed, the GetCurrentClusterMetadata() would load for the change cluster name, so the returned response's ClusterName would always equal to the currentClusterName from config.

File: temporal/fx.go [hunk 4]
```diff
@@ -481,46 +481,94 @@ func ApplyClusterMetadataConfigProvider(
...
+		if clusterInfo.InitialFailoverVersion != resp.InitialFailoverVersion {
+			logger.Error(
+				mismatchLogMessage,
+				tag.Key("clusterInfo.initialFailoverVersion"),
+				tag.IgnoredValue(clusterInfo.InitialFailoverVersion),
+				tag.Value(resp.InitialFailoverVersion))
+
+			clusterInfo.InitialFailoverVersion = resp.InitialFailoverVersion
+		}
+		if clusterInfo.Enabled != resp.IsConnectionEnabled {
```
Comment (yiminc) [inline temporal/fx.go:561]: should check resp.IsConnectionEnabled first and skip all validation if enabled is false.

File: common/persistence/cassandra/cluster_metadata_store.go (hunk 1 above)
Comment (yux0) [reply common/persistence/cassandra/cluster_metadata_store.go:141]: Delete current cluster is not allowed. This handles in the clusterMetadataStore.go in L165

File: temporal/fx.go (hunk 2 above)
Comment (yux0) [reply temporal/fx.go:497]: Yes. We unset this field in L486. With another question, does the remote cluster Id matter ?

File: temporal/fx.go (hunk 4 above)
Comment (yux0) [reply temporal/fx.go:561]: Make sense. Will update it

File: temporal/fx.go (hunk 3 above)
Comment (yux0) [reply temporal/fx.go:521]: Correct. This check is pointless with the new change. I am going to remove it.

File: common/persistence/clusterMetadataStore.go [hunk 5]
```diff
@@ -131,12 +154,26 @@ func (m *clusterMetadataManagerImpl) SaveClusterMetadata(request *SaveClusterMet
...
+	return m.persistence.SaveClusterMetadata(&InternalSaveClusterMetadataRequest{
+		ClusterName:     request.ClusterName,
+		ClusterMetadata: mcm,
+		Version:         request.Version,
+	})
+}
+
+func (m *clusterMetadataManagerImpl) DeleteClusterMetadata(request *DeleteClusterMetadataRequest) error {
+	if request.ClusterName == m.currentClusterName {
+		return serviceerror.NewInvalidArgument("Cannot delete cluster metadata")
```
Comment (yiminc) [inline common/persistence/clusterMetadataStore.go:166]: message should be cannot delete current cluster metadata

**Review by yiminc — APPROVED** [review]

File: common/persistence/clusterMetadataStore.go (hunk 5 above)
Comment (yux0) [reply common/persistence/clusterMetadataStore.go]: updated the error message

**Pull request merged**